src/BSON/Document.c

```diff
@@ -0,0 +1,355 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <php.h>
+#include <ext/standard/base64.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "phongo_bson_encode.h"
+#include "BSON/Document_arginfo.h"
+#include "BSON/Iterator.h"
+
+zend_class_entry* php_phongo_document_ce;
+
+/* Initialize the object and return whether it was successful. An exception will
+ * be thrown on error. */
+static bool php_phongo_document_init(php_phongo_document_t* intern, bson_t* bson)
+{
+	intern->bson = bson;
+
+	return true;
+}
+
+static HashTable* php_phongo_document_get_properties_hash(phongo_compat_object_handler_type* object, bool is_temp)
+{
+	php_phongo_document_t* intern;
+	HashTable*             props;
+
+	intern = Z_OBJ_DOCUMENT(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_temp, intern, props, 2);
+
+	if (!intern->bson) {
+		return props;
+	}
+
+	{
+		zval data, length;
+
+		ZVAL_STRINGL(&data, (const char*) bson_get_data(intern->bson), intern->bson->len);
+		zend_hash_str_update(props, "data", sizeof("data") - 1, &data);
+
+		ZVAL_LONG(&length, intern->bson->len);
+		zend_hash_str_update(props, "length", sizeof("length") - 1, &length);
+	}
+
+	return props;
+}
+
+PHONGO_DISABLED_CONSTRUCTOR(MongoDB_BSON_Document)
+
+static PHP_METHOD(MongoDB_BSON_Document, fromBSONString)
+{
+	zval                   zv;
+	php_phongo_document_t* intern = NULL;
+	zend_string*           bson_string;
+	const bson_t*          bson;
+	bson_reader_t*         reader;
+	bool                   eof = false;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_STR(bson_string)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	reader = bson_reader_new_from_data((const uint8_t*) ZSTR_VAL(bson_string), ZSTR_LEN(bson_string));
+	if (!(bson = bson_reader_read(reader, NULL))) {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Could not read document from BSON reader");
+
+		goto cleanup;
+	}
+
+	object_init_ex(&zv, php_phongo_document_ce);
+	intern       = Z_DOCUMENT_OBJ_P(&zv);
+	intern->bson = bson_copy(bson);
+
+	if (bson_reader_read(reader, &eof) || !eof) {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Reading document did not exhaust input buffer");
+		zval_ptr_dtor(&zv);
+
+		goto cleanup;
+	}
+
+	RETVAL_ZVAL(&zv, 1, 1);
+
+cleanup:
+	if (reader) {
+		bson_reader_destroy(reader);
+	}
+}
+
+static PHP_METHOD(MongoDB_BSON_Document, fromJSON)
+{
+	zval                   zv;
+	php_phongo_document_t* intern;
+	zend_string*           json;
+	bson_t*                bson;
+	bson_error_t           error;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_STR(json)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	bson = bson_new_from_json((const uint8_t*) ZSTR_VAL(json), ZSTR_LEN(json), &error);
```

Comment:
I just realized that `bson_new_from_json` and `bson_init_from_json` do not validate that the JSON string contains exactly one document. Reported in [CDRIVER-4533](https://jira.mongodb.org/browse/CDRIVER-4533).

Not sure if that's something you care to check here. I don't think our existing `fromJSON` function cares, as it's also a thin wrapper around this function. The [`fromJSON()` docs](https://www.php.net/manual/en/function.mongodb.bson-fromjson.php) also don't talk about whether extra data is permitted.

Reply:
I wouldn't check that here - this class is merely a nice place for those functions to live. I agree with this check happening in libbson though.